op-program/client/interop/interop.go

```diff
@@ -0,0 +1,115 @@
+package interop
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-program/client/boot"
+	"github.com/ethereum-optimism/optimism/op-program/client/claim"
+	"github.com/ethereum-optimism/optimism/op-program/client/interop/types"
+	"github.com/ethereum-optimism/optimism/op-program/client/l1"
+	"github.com/ethereum-optimism/optimism/op-program/client/l2"
+	"github.com/ethereum-optimism/optimism/op-program/client/tasks"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+var (
+	ErrIncorrectOutputRootType = errors.New("incorrect output root type")
+)
+
+type taskExecutor interface {
+	RunDerivation(
+		logger log.Logger,
+		rollupCfg *rollup.Config,
+		l2ChainConfig *params.ChainConfig,
+		l1Head common.Hash,
+		agreedOutputRoot eth.Bytes32,
+		claimedBlockNumber uint64,
+		l1Oracle l1.Oracle,
+		l2Oracle l2.Oracle) (tasks.DerivationResult, error)
+}
+
+func RunInteropProgram(logger log.Logger, bootInfo *boot.BootInfo, l1PreimageOracle l1.Oracle, l2PreimageOracle l2.Oracle) error {
+	return runInteropProgram(logger, bootInfo, l1PreimageOracle, l2PreimageOracle, &interopTaskExecutor{})
+}
+
+func runInteropProgram(logger log.Logger, bootInfo *boot.BootInfo, l1PreimageOracle l1.Oracle, l2PreimageOracle l2.Oracle, tasks taskExecutor) error {
+	logger.Info("Interop Program Bootstrapped", "bootInfo", bootInfo)
+
+	// For the first step in a timestamp, we would get a SuperRoot as the agreed claim - TransitionStateByRoot will
+	// automatically convert it to a TransitionState with Step: 0.
+	transitionState := l2PreimageOracle.TransitionStateByRoot(bootInfo.L2OutputRoot)
+	if transitionState.Version() != types.IntermediateTransitionVersion {
+		return fmt.Errorf("%w: %v", ErrIncorrectOutputRootType, transitionState.Version())
+	}
+
+	super, err := eth.UnmarshalSuperRoot(transitionState.SuperRoot)
+	if err != nil {
+		return fmt.Errorf("invalid super root: %w", err)
+	}
+	if super.Version() != eth.SuperRootVersionV1 {
+		return fmt.Errorf("%w: %v", ErrIncorrectOutputRootType, super.Version())
+	}
+	superRoot := super.(*eth.SuperV1)
+	claimedBlockNumber, err := bootInfo.RollupConfig.TargetBlockNumber(superRoot.Timestamp + 1)
```

Comment:
Assuming a 2 second block time and the superRoot.Timestamp is at a block time boundary, then the claimed block remains unchanged. So the client only needs to check that the disputed output matches the agreed output. Worth implementing the simple check as an optimization?

Reply:
I think the derivation will skip out immediately for us anyway. It is worth having an action test for and checking it all happens quickly though. Will follow up on that - we need to test the no-op steps as well which aren't yet implemented.